crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2615,36 +2615,112 @@ impl<'db> Type<'db> {
+                        },
+                    )))
+                } else {
+                    // User defined class
+                    let instance_ty = Type::Instance(InstanceType { class });
+
+                    // Try calling __init__ to check arguments before returning the instance type
+                    match instance_ty.try_call_dunder(db, "__init__", arguments) {
+                        Err(CallDunderError::Call(mut error)) => {
+                            // If the call to `__init__` fails we want to return inner error directly
+                            // but ensure that the fallback return type is the instance type.
+
+                            match error {
+                                CallError::NotCallable { .. } => {
+                                    // Instead of trying to emit a diagnostic at call site, these should be
+                                    // caught at the definition site of the class. Here we return Ok with the
+                                    // instance type to ensure that the return type is correct.
+                                    Ok(CallOutcome::Single(CallBinding::from_return_type(
+                                        instance_ty,
+                                    )))
+                                }
+                                CallError::PossiblyUnboundDunderCall { .. } => {
+                                    tracing::warn!("Inferred `__init__` as possibly unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
+
+                                    // We return None (see the logic for `NotCallable` above).
+                                    Ok(CallOutcome::Single(CallBinding::from_return_type(
+                                        instance_ty,
+                                    )))
+                                }
+                                CallError::Union(_) | CallError::BindingError { .. } => {
+                                    // Replace the return type and return as Err
+                                    error.set_return_type(instance_ty);
+                                    Err(error)
+                                }
+                            }
+                        }
+                        // If we are using vendored typeshed, it should be impossible to have missing
+                        // or unbound `__init__` methods on classes, as all classes have `object` in MRO,
+                        // hence we emit user warnings for these cases.
+                        Err(CallDunderError::PossiblyUnbound(mut outcome)) => {
+                            tracing::warn!("Inferred `__init__` as possibly unbound on a class. All classes have `object` in MRO, so this is not possible. This indicates an error in typeshed.");
+
+                            outcome.set_return_type(instance_ty);
+                            Ok(outcome)
```

Comment:
Technically we _could_ explicitly also check against a single argument call here, and union the outcome with the other outcome, which would make this fallback "perfect". But that also seems like a hassle to implement for a case that will hardly ever occur. I'd also be fine saying if you are missing `object.__init__` in typeshed you just don't get your calls checked against it.